En este archivo puedes escribir lo que estimes conveniente. Te recomendamos detallar tu solución y todas las suposiciones que estás considerando. Aquí puedes ejecutar las funciones que definiste en los otros archivos de la carpeta src, medir el tiempo, memoria, etc.

In [1]:
file_path = "farmers-protest-tweets-2021-2-4.json"

In [2]:
import cProfile
import pstats

%load_ext memory_profiler

# Q1

## Q1 Time

In [3]:
from q1_time import q1_time

file_path = "farmers-protest-tweets-2021-2-4.json"
with cProfile.Profile() as profile:
    result_q1_time = q1_time(file_path)

result_q1_time

[(datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 17), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 16), 'jot__b'),
 (datetime.date(2021, 2, 14), 'rebelpacifist'),
 (datetime.date(2021, 2, 18), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 15), 'jot__b'),
 (datetime.date(2021, 2, 20), 'MangalJ23056160'),
 (datetime.date(2021, 2, 23), 'Surrypuria'),
 (datetime.date(2021, 2, 19), 'Preetm91')]

Profiling

In [4]:
results = pstats.Stats(profile)
results.sort_stats(pstats.SortKey.TIME)
results.print_stats()

         1643269 function calls (1642622 primitive calls) in 11.566 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   117407    5.402    0.000    5.402    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\decoder.py:343(raw_decode)
        1    2.539    2.539   11.566   11.566 c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q1_time.py:6(q1_time)
    49772    1.723    0.000    1.723    0.000 {built-in method _codecs.charmap_decode}
   117407    0.399    0.000    6.138    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\decoder.py:332(decode)
        1    0.311    0.311    0.317    0.317 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\core\tools\datetimes.py:456(_array_strptime_with_fallback)
   117407    0.267    0.000    6.493    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\__init__.py:299(lo

El tiempo de ejecución es de 11.5 segundos.

In [5]:
%%mprun -f  q1_time
q1_time(file_path)

Filename: c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q1_time.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     6    124.9 MiB    124.9 MiB           1   def q1_time(file_path: str) -> List[Tuple[datetime.date, str]]:
     7                                         
     8    124.9 MiB      0.0 MiB           1       dict_data = {}
     9    124.9 MiB      0.0 MiB           1       ids = []
    10    124.9 MiB      0.0 MiB           1       dates = []
    11    124.9 MiB      0.0 MiB           1       usernames = []
    12    140.6 MiB      0.0 MiB           2       with open(file_path) as json_file:
    13    140.6 MiB  -1890.8 MiB      117408           for line in json_file:
    14    140.6 MiB  -1880.1 MiB      117407               json_obj = json.loads(line)
    15    140.6 MiB  -1890.2 MiB      117407               ids.append(json_obj['id'])
    16    140.6 MiB  -1902.5 MiB      117407               dates.append(json_obj['d

El máximo uso de memoria de 150 MiB.

## Q1 Memory

In [6]:
from q1_memory import q1_memory

file_path = "farmers-protest-tweets-2021-2-4.json"
with cProfile.Profile() as profile:
    result_q1_memory = q1_memory(file_path)

result_q1_memory

[(datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 17), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 16), 'jot__b'),
 (datetime.date(2021, 2, 14), 'rebelpacifist'),
 (datetime.date(2021, 2, 18), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 15), 'jot__b'),
 (datetime.date(2021, 2, 20), 'MangalJ23056160'),
 (datetime.date(2021, 2, 23), 'Surrypuria'),
 (datetime.date(2021, 2, 19), 'Preetm91')]

Profiling

In [7]:
results = pstats.Stats(profile)
results.sort_stats(pstats.SortKey.TIME)
results.print_stats()

         1627126 function calls (1626952 primitive calls) in 11.368 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   117407    5.330    0.000    5.330    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\decoder.py:343(raw_decode)
        1    2.533    2.533   11.368   11.368 c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q1_memory.py:6(q1_memory)
    49772    1.797    0.000    1.797    0.000 {built-in method _codecs.charmap_decode}
   117407    0.399    0.000    6.055    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\decoder.py:332(decode)
       12    0.351    0.029    0.351    0.029 {method 'collect' of 'builtins.PyLazyFrame' objects}
   117407    0.266    0.000    6.409    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\__init__.py:299(loads)
   234814    0.240    0.000    0.240    0.000 {method 'match' of 're.Pattern'

El tiempo de ejecución es de 11.3 segundos.

In [8]:
%%mprun -f  q1_memory
q1_memory(file_path)

Filename: c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q1_memory.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     6    162.0 MiB    162.0 MiB           1   def q1_memory(file_path: str) -> List[Tuple[datetime.date, str]]:
     7                                             
     8    162.0 MiB      0.0 MiB           1       dict_data = {}
     9    162.0 MiB      0.0 MiB           1       ids = []
    10    162.0 MiB      0.0 MiB           1       dates = []
    11    162.0 MiB      0.0 MiB           1       usernames = []
    12    175.0 MiB      0.0 MiB           2       with open(file_path) as json_file:
    13    175.0 MiB    -14.3 MiB      117408           for line in json_file:
    14    175.0 MiB     -8.1 MiB      117407               json_obj = json.loads(line)
    15    175.0 MiB    -11.4 MiB      117407               ids.append(json_obj['id'])
    16    175.0 MiB  -1475.0 MiB      117407               dates.append(jso

El máximo uso de memoria es de 180 MiB.

# Q2

## Q2 Time

In [9]:
from q2_time import q2_time

file_path = "farmers-protest-tweets-2021-2-4.json"
with cProfile.Profile() as profile:
    result_q2_time = q2_time(file_path)

result_q2_time

[('🙏', 5049),
 ('😂', 3072),
 ('🚜', 2972),
 ('🌾', 2182),
 ('🇮🇳', 2086),
 ('🤣', 1668),
 ('✊', 1651),
 ('❤️', 1382),
 ('🙏🏻', 1317),
 ('💚', 1040)]

Profiling

In [10]:
results = pstats.Stats(profile)
results.sort_stats(pstats.SortKey.TIME)
results.print_stats()

         87496064 function calls (87496029 primitive calls) in 70.868 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 17140706   27.903    0.000   42.665    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\site-packages\emoji\tokenizer.py:158(tokenize)
   117407   13.879    0.000   59.845    0.001 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\site-packages\emoji\core.py:366(<listcomp>)
 17023302    7.165    0.000   11.858    0.000 <string>:1(<lambda>)
   117407    5.152    0.000    5.152    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\decoder.py:343(raw_decode)
 17023311    4.693    0.000    4.693    0.000 {built-in method __new__ of type object at 0x00007FFAB536A920}
17141014/17141013    3.326    0.000    3.326    0.000 {built-in method builtins.isinstance}
 17140712    2.805    0.000    2.805    0.000 {method 'append' of 'list' objects}
        1    2.314    2.314   7

El tiempo de ejecución es de 70 segundos.

In [11]:
%%mprun -f  q2_time
q2_time(file_path)

Filename: c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q2_time.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     6    179.3 MiB    179.3 MiB           1   def q2_time(file_path: str) -> List[Tuple[str, int]]:
     7                                         
     8    179.3 MiB      0.0 MiB           1       content_list = []
     9    200.0 MiB      0.0 MiB           2       with open(file_path) as json_file:
    10    200.0 MiB     -2.2 MiB      117408           for line in json_file:
    11    200.0 MiB     -4.7 MiB      117407               json_obj = json.loads(line)
    12    200.0 MiB     -3.2 MiB      117407               content_list.append(json_obj['content'])
    13                                             
    14    199.6 MiB     -0.4 MiB           1       content = pd.Series(content_list, name='content')
    15    204.5 MiB      4.9 MiB           1       emojis = content.apply(get_emojis_in_content)
    16    206.

El máximo uso de memoria es de 206 MiB.

## Q2 Memory

In [12]:
from q2_memory import q2_memory

file_path = "farmers-protest-tweets-2021-2-4.json"
with cProfile.Profile() as profile:
    result_q2_memory = q2_memory(file_path)

result_q2_memory

[('🙏', 5049),
 ('😂', 3072),
 ('🚜', 2972),
 ('🌾', 2182),
 ('🇮🇳', 2086),
 ('🤣', 1668),
 ('✊', 1651),
 ('❤️', 1382),
 ('🙏🏻', 1317),
 ('💚', 1040)]

### Profiling

In [13]:
results = pstats.Stats(profile)
results.sort_stats(pstats.SortKey.TIME)
results.print_stats()

         87608079 function calls (87608029 primitive calls) in 70.661 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 17140706   28.632    0.000   43.885    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\site-packages\emoji\tokenizer.py:158(tokenize)
   117407   14.178    0.000   61.311    0.001 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\site-packages\emoji\core.py:366(<listcomp>)
 17023302    7.316    0.000   12.327    0.000 <string>:1(<lambda>)
 17023302    5.012    0.000    5.012    0.000 {built-in method __new__ of type object at 0x00007FFAB536A920}
   117407    4.220    0.000    4.220    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\decoder.py:343(raw_decode)
 17140707    3.269    0.000    3.269    0.000 {built-in method builtins.isinstance}
 17140712    2.829    0.000    2.829    0.000 {method 'append' of 'list' objects}
        1    2.031    2.031   70.661   

El tiempo de ejecución es de 70 segundos.

In [14]:
%%mprun -f  q2_memory
q2_memory(file_path)

Filename: c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q2_memory.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     7    181.3 MiB    181.3 MiB           1   def q2_memory(file_path: str) -> List[Tuple[str, int]]:
     8                                             
     9    181.3 MiB      0.0 MiB           1       dict_data = {}
    10    181.3 MiB      0.0 MiB           1       content_list = []
    11    202.6 MiB      0.0 MiB           2       with open(file_path) as json_file:
    12    202.6 MiB      2.0 MiB      117408           for line in json_file:
    13    202.6 MiB     -0.3 MiB      117407               json_obj = json.loads(line)
    14    202.6 MiB      1.6 MiB      117407               content_list.append(json_obj['content'])
    15                                         
    16    202.6 MiB      0.0 MiB           1       dict_data['content'] = content_list
    17                                         
    18 

El máximo uso de memoria es de 202 MiB.

# Q3

## Q3 Time

In [15]:
from q3_time import q3_time

file_path = "farmers-protest-tweets-2021-2-4.json"
with cProfile.Profile() as profile:
    result_q3_time = q3_time(file_path)

result_q3_time 

[('narendramodi', 2261),
 ('Kisanektamorcha', 1836),
 ('RakeshTikaitBKU', 1639),
 ('PMOIndia', 1422),
 ('RahulGandhi', 1125),
 ('GretaThunberg', 1046),
 ('RaviSinghKA', 1015),
 ('rihanna', 972),
 ('UNHumanRights', 962),
 ('meenaharris', 925)]

Profiling

In [16]:
results = pstats.Stats(profile)
results.sort_stats(pstats.SortKey.TIME)
results.print_stats()

         1392598 function calls (1392554 primitive calls) in 9.173 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   117407    4.445    0.000    4.445    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\decoder.py:343(raw_decode)
        1    1.979    1.979    9.173    9.173 c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q3_time.py:6(q3_time)
    49772    1.532    0.000    1.532    0.000 {built-in method _codecs.charmap_decode}
   117407    0.321    0.000    5.028    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\decoder.py:332(decode)
   117407    0.216    0.000    5.312    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\__init__.py:299(loads)
   234814    0.194    0.000    0.194    0.000 {method 'match' of 're.Pattern' objects}
        1    0.176    0.176    0.187    0.187 c:\Users\migue\AppData\Local\Programs\Python\Pyt

El tiempo de ejecución es de 9.1 segundos.

In [17]:
%%mprun -f  q3_time
q3_time(file_path)

Filename: c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q3_time.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     6    206.3 MiB    206.3 MiB           1   def q3_time(file_path: str) -> List[Tuple[str, int]]:
     7                                         
     8    206.3 MiB      0.0 MiB           1       content_list = []
     9    240.4 MiB      0.0 MiB           2       with open(file_path) as json_file:
    10    240.4 MiB      3.1 MiB      117408           for line in json_file:
    11    240.4 MiB     20.9 MiB      117407               json_obj = json.loads(line)
    12    240.4 MiB      7.3 MiB      117407               content_list.append(json_obj['content'])
    13                                         
    14    242.2 MiB      1.8 MiB           1       content = pd.Series(content_list, name='content')
    15    229.8 MiB    -12.4 MiB           1       mentioned_users = content.str.findall(('@(\w+)'))
    16    231.

El máximo uso de memoria es de 242 MiB.

## Q3 Memory

In [18]:
from q3_memory import q3_memory

file_path = "farmers-protest-tweets-2021-2-4.json"
with cProfile.Profile() as profile:
    result_q3_memory = q3_memory(file_path)

result_q3_memory

[('narendramodi', 2261),
 ('Kisanektamorcha', 1836),
 ('RakeshTikaitBKU', 1639),
 ('PMOIndia', 1422),
 ('RahulGandhi', 1125),
 ('GretaThunberg', 1046),
 ('RaviSinghKA', 1015),
 ('rihanna', 972),
 ('UNHumanRights', 962),
 ('meenaharris', 925)]

In [19]:
results = pstats.Stats(profile)
results.sort_stats(pstats.SortKey.TIME)
results.print_stats()

         2228560 function calls in 9.078 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   117407    4.249    0.000    4.249    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\decoder.py:343(raw_decode)
        1    1.903    1.903    9.078    9.078 c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q3_memory.py:11(q3_memory)
    49772    1.483    0.000    1.483    0.000 {built-in method _codecs.charmap_decode}
   117407    0.308    0.000    4.808    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\decoder.py:332(decode)
   117407    0.207    0.000    5.080    0.000 c:\Users\migue\AppData\Local\Programs\Python\Python310\lib\json\__init__.py:299(loads)
   234814    0.186    0.000    0.186    0.000 {method 'match' of 're.Pattern' objects}
   117407    0.173    0.000    0.173    0.000 {method 'findall' of 're.Pattern' objects}
   117407    0.086    0.00

El tiempo de ejecución es de 9.1 segundos.

In [20]:
%%mprun -f  q3_memory
q3_memory(file_path)

Filename: c:\Users\migue\Documents\Búsqueda de Trabajo\Entrevistas\Latam\challenge_DE_01\src\q3_memory.py

Line #    Mem usage    Increment  Occurrences   Line Contents
    11    189.4 MiB    189.4 MiB           1   def q3_memory(file_path: str) -> List[Tuple[str, int]]:
    12                                             
    13    189.4 MiB      0.0 MiB           1       dict_data = {}
    14    189.4 MiB      0.0 MiB           1       content_list = []
    15    209.7 MiB     -0.0 MiB           2       with open(file_path) as json_file:
    16    209.7 MiB   -108.4 MiB      117408           for line in json_file:
    17    209.7 MiB    -79.4 MiB      117407               json_obj = json.loads(line)
    18    209.7 MiB   -108.1 MiB      117407               content_list.append(json_obj['content'])
    19                                         
    20    209.7 MiB      0.0 MiB           1       dict_data['content'] = content_list
    21    212.7 MiB      2.9 MiB           1       ment

El máximo uso de memoria es de 213 MiB.